File: src/main/python/litcoder/contests/contest-1/MaximumAlternatingSubsequenceSum.py
"""Solution to Maximum Alternating Subsequence Sum"""

from typing import List


def solution(arr: List[int]) -> int:
    """Returns maximum alternating sum of the array

    Args:
        arr (List[int]): array of numbers

    Returns:
        int: max alternating sum
    """
    max_sum = 0

    subsequences = get_all_subsequences(arr)
    for subsequence in subsequences:
        even_indices_sum = get_even_indices_sum(subsequence)
        odd_indices_sum = get_odd_indices_sum(subsequence)

        alternating_sum = even_indices_sum - odd_indices_sum

        max_sum = max(max_sum, alternating_sum)

    return max_sum


def get_even_indices_sum(arr: List[int]) -> int:
    """Returns  sum of numbers present at even indices

    Args:
        arr (List[int]): array of numbers

    Returns:
        int: sum
    """
    if len(arr) == 0:
        return 0
    return sum([i for i in arr[::2]])


def get_odd_indices_sum(arr: List[int]) -> int:
    """Returns sum of numbers present at odd indices

    Args:
        arr (List[int]): array of numbers

    Returns:
        int: sum
    """
    if len(arr) == 0:
        return 0
    return sum([i for i in arr[1::2]])


def get_subsequences(
    arr: List[int], start: int, curr_seq: List[int], result: List[List[int]]
) -> None:
    """Generates all possible subsequences for the given array starting from
    given index and stores the result in result array

    Args:
        arr (List[int]): array of numbers
        start (int): index to start at
        curr_seq (List[int]): current subsequence, meant to be used inside recur
        sion
        result (List[List[int]]): list of all subsequences
    """
    result.append(list(curr_seq))

    for i in range(start, len(arr)):
        if i > start and arr[i] == arr[i - 1]:
            continue

        curr_seq.append(arr[i])
        get_subsequences(arr, i + 1, curr_seq, result)

        curr_seq.pop()


def get_all_subsequences(arr: List[int]) -> List[List[int]]:
    """Wrapper function for get_subsequences to call it easily and get result
    without passing unnecessary arguments

    Args:
        arr (List[int]): array of numbers

    Returns:
        List[List[int]]: list of all subsequences
    """
    result: List[List[int]] = []
    curr_seq: List[int] = []

    get_subsequences(arr, 0, curr_seq, result)
    return result


testcases = int(input())
for _ in range(testcases):
    arr = list(map(int, input().split(" ")))
    print(solution(arr))

File: src/codes/college/litcoder/contests/contest1/MaximumAlternatingSubsequenceSum.java
package codes.college.litcoder.contests.contest1;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class MaximumAlternatingSubsequenceSum {

    /**
     * Calculates the maximum alternating subsequence sum from the given list of
     * integers.
     *
     * @param arr The list of integers
     * @return The maximum alternating subsequence sum
     */
    public static int solution(List<Integer> arr) {
        int maxSum = 0;

        List<List<Integer>> subsequences = getAllSubsequences(arr);
        for (List<Integer> subsequence : subsequences) {
            int evenIndicesSum = getEvenIndicesSum(subsequence);
            int oddIndicesSum = getOddIndicesSum(subsequence);

            int alternatingSum = evenIndicesSum - oddIndicesSum;

            maxSum = Math.max(maxSum, alternatingSum);
        }

        return maxSum;
    }

    /**
     * Calculates the sum of elements at even indices in the given list.
     *
     * @param arr the list of integers
     * @return the sum of elements at even indices
     */
    public static int getEvenIndicesSum(List<Integer> arr) {
        if (arr.isEmpty()) {
            return 0;
        }
        return arr
                .stream()
                .filter(i -> arr.indexOf(i) % 2 == 0)
                .mapToInt(Integer::intValue)
                .sum();
    }

    /**
     * Calculates the sum of elements at odd indices in the given list.
     *
     * @param arr the list of integers
     * @return the sum of elements at odd indices
     */
    public static int getOddIndicesSum(List<Integer> arr) {
        if (arr.isEmpty()) {
            return 0;
        }
        return arr
                .stream()
                .filter(i -> arr.indexOf(i) % 2 != 0)
                .mapToInt(Integer::intValue)
                .sum();
    }

    /**
     * Generates all possible subsequences of the given array starting from the
     * specified index.
     * Each subsequence is added to the result list.
     *
     * @param arr     the input array
     * @param start   the starting index
     * @param currSeq the current subsequence being generated
     * @param result  the list to store the generated subsequences
     */
    public static void getSubsequences(
            List<Integer> arr,
            int start,
            List<Integer> currSeq,
            List<List<Integer>> result) {
        result.add(new ArrayList<>(currSeq));

        for (int i = start; i < arr.size(); i++) {
            if (i > start && arr.get(i).equals(arr.get(i - 1))) {
                continue;
            }

            currSeq.add(arr.get(i));
            getSubsequences(arr, i + 1, currSeq, result);

            currSeq.removeLast();
        }
    }

    /**
     * Returns a list of all subsequences of the given array.
     *
     * @param arr the input array
     * @return a list of all subsequences
     */
    public static List<List<Integer>> getAllSubsequences(List<Integer> arr) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> currSeq = new ArrayList<>();

        getSubsequences(arr, 0, currSeq, result);
        return result;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int arrSize = scanner.nextInt();
        List<Integer> arr = new ArrayList<>();
        for (int i = 0; i < arrSize; i++) {
            arr.add(scanner.nextInt());
        }

        System.out.println(solution(arr));

        scanner.close();
    }
}
